### PR TITLE
Fix makewrapper unsupported args

### PR DESCRIPTION
**Pull request by ahmedtd**
###### Motivation for this change

Previously, makeWrapper would accept arguments it didn't recognize,
potentially allowing argument misspellings or broken callers.

Now, makeWrapper dies with a backtrace if it is called incorrectly.

Also changes `wrapProgram` so that it doesn't pass through the first
argument twice --- this was tripping up the argument checking.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by ahmedtd** [comment]
PR24944 depends on this, for the `die` utility function

**Comment by Ericson2314** [comment]
Is this a mass rebuild? I'd think so.

**Comment by ahmedtd** [comment]
Yes.  This PR was split from [https://github.com/NixOS/nixpkgs/pull/24944](), which might provide more context.

Currently, `nox-review wip --against master` chokes on it, since `nixpkgs` includes some problematic attribute names, which this change pulls in.  See #25074 and NixOS/nix#1342 for the discussions.

I'm not sure how that will end up.

**Comment by Ericson2314** [comment]
@ahmedtd Ok that's cool. I guess I don't know whether the policy is to submit known mass-rebuilds to staging or master.

**Comment by ahmedtd** [comment]
Oh, I thought commits went to staging after passing Hydra on master.  Is
there some documentation that lays out which branches are what?

**Comment by Ericson2314** [comment]
I'm not sure, but the point of staging is to queue up mass rebuilds so they cancel out a bit for master. There is something about not just merging any old untested thing to staging, but I think that means test locally. This stuff indeed ought to be documented better :).

**Comment by joachifm** [comment]
Regarding commit policy, see https://nixos.org/nixpkgs/manual/#idm140737317211248

**Comment by FRidh** [comment]
I've changed the target branch to staging.

File: pkgs/build-support/setup-hooks/make-wrapper.sh [hunk 1]
```diff
@@ -32,26 +32,20 @@ makeWrapper() {
     for ((n = 2; n < ${#params[*]}; n += 1)); do
         p="${params[$n]}"
 
-        if test "$p" = "--set"; then
+        if [[ "$p" == "--set" ]]; then
```
Comment (zimbatm) [inline pkgs/build-support/setup-hooks/make-wrapper.sh]: wouldn't it make sense to use a switch case here instead?

File: pkgs/build-support/setup-hooks/die.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    # Print a backtrace.
+    local frame=0
+    while caller $frame; do
+        ((frame++));
+    done
+    printf "\n"
+
+    exit 1
+}
```
Comment (zimbatm) [inline pkgs/build-support/setup-hooks/die.sh]: How does it look like in practice when that happens? Nix's bash is quite convoluted and might have some weird stack trace.

File: pkgs/build-support/setup-hooks/make-wrapper.sh (hunk 1 above)
Comment (ahmedtd) [inline pkgs/build-support/setup-hooks/make-wrapper.sh]: A previous PR that did something similar (#19061) ran up against the OSX system version of bash being too old to support the case statement.

File: pkgs/build-support/setup-hooks/die.sh (hunk 2 above)
Comment (ahmedtd) [inline pkgs/build-support/setup-hooks/die.sh]: It varies in usefulness, depending on how makeWrapper is being called.  Sometimes there's enough information to be helpful, but often the stack trace stops at the point where `stdenv/setup.sh` evaluates the current phase hook.

File: pkgs/build-support/setup-hooks/make-wrapper.sh (hunk 1 above)
Comment (zimbatm) [inline pkgs/build-support/setup-hooks/make-wrapper.sh]: fair, the fallthrough is only useful for a single pair of arguments which could be worked around: https://github.com/NixOS/nixpkgs/pull/19061#discussion_r113685418

**Comment by peti** [comment]
@ahmedtd, this PR has mere conflicts. Would you mind force-pushing an update that is re-base relative to current versions of `staging`?

File: pkgs/build-support/setup-hooks/make-wrapper.sh (hunk 1 above)
Comment (Ericson2314) [inline pkgs/build-support/setup-hooks/make-wrapper.sh]: Ok Bash 3.2 is not *that* bad. The issue was not case but fallthrough, and or-patterns can be used instead (and in fact are a better solution).
```bash
# Works in bash 3.2
case n in
    p|n) echo yup;;
esac
```

**Comment by ahmedtd** [comment]
Rebased onto current staging.

**Comment by FRidh** [comment]
This breaks several packages:
```
installing

Builder called die: makeWrapper doesn't understand the arg /homeless-shelter
Backtrace:
88 makeWrapper /nix/store/ixm1iljz3bgl04z4c2k2nvmq9wian9qh-hook/nix-support/setup-hook
129 wrapProgram /nix/store/ixm1iljz3bgl04z4c2k2nvmq9wian9qh-hook/nix-support/setup-hook
987 genericBuild /nix/store/0ifjmsk5s6nqxz5wlkb2c78xf1213dkc-stdenv/setup
2 main /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh

builder for ‘/nix/store/67l8nyc4b7rbyld7clby9lkibraphn9q-nix-prefetch-git.drv’ failed with exit code 1
```
https://hydra.nixos.org/build/58313566#tabs-summary

**Comment by Ericson2314** [comment]
While we're at it, if someone could use use `case` with `--suffix-contents | --prefix-contents)` alternatives, per https://github.com/NixOS/nixpkgs/pull/25185/files#r129576600?

**Comment by ahmedtd** [comment]
This looks like preexisting brokenness in nix-prefetch-scripts.  I can make
a pr to fix it tonight.

**Comment by globin** [comment]
@ahmedtd fixed already

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/compare/staging...obsidiansystems:make-wrapper-clean Here is a draft of that cleanup, if anyone wants to finish it.

**Comment by FRidh** [comment]
Just a note for a future improvement or someone else that encounters this. 

The command `die()` is not accessible when using `buildEnv`. Therefore, when you wrap things from `buildEnv` you may get `nix-support/setup-hook: line 7: die: command not found`.

**Comment by ahmedtd** [comment]
@FRidh That will be a really frustrating experience to debug --- I will send a PR to replace `die` with its contents in `make-wrapper.sh`

**Comment by ahmedtd** [comment]
@FRidh Would you mind telling me in which file you observed this problem?  Both `die.sh` and `make-wrapper.sh` are setup hooks, so I don't understand how one could be accessible but not the other.

**Comment by globin** [comment]
@ahmedtd this happens when using `python.withPackages` prior to 4495bfe1382ea779f02996d0e18d25d892678e8c

=> `buildEnv` with a bad `makeWrapper` in `postInstall` should make this reproducable.

**Comment by FRidh** [comment]
Indeed. An example that has scripts in `/bin` is `python.withPackages(ps: [ps.sphinx])`.

**Pull request merged**